server/src/lib/model/db/migrations/20191024074951-challenge-related-schemas.ts

```diff
@@ -0,0 +1,73 @@
+export const up = async function(db: any): Promise<any> {
+  return db.runSql(
+    `
+    /* Keep user_clients clean */
+    ALTER TABLE user_clients DROP FOREIGN KEY uc_ibfk_1, DROP COLUMN challenge_team;
+
+    /* A list of challenges. This will only hold one row (the pilot challenge) until at least 2020. */
+    CREATE TABLE challenges (
+      id INT UNSIGNED NOT NULL AUTO_INCREMENT PRIMARY KEY,
+
+      name VARCHAR(255),
+      url_token TEXT NOT NULL,
+      start_date DATETIME NOT NULL
+    );
+
+    ALTER TABLE teams DROP PRIMARY KEY,
+              ADD COLUMN id INT UNSIGNED NOT NULL AUTO_INCREMENT PRIMARY KEY, /* better performance and shorter sql statements */
+              CHANGE invite_token url_token CHAR(36) NOT NULL,
+              CHANGE team_name name VARCHAR(255),
+              ADD COLUMN challenge_id INT UNSIGNED NOT NULL,                  /* teams has to be associated with exact one challenge to make sense */
+              ADD FOREIGN KEY (challenge_id) REFERENCES challenges(id);
+              
+    /* Records each enrollment (a user signing up for a team) across all challenges. */
+    CREATE TABLE enroll (
+      id INT UNSIGNED NOT NULL AUTO_INCREMENT PRIMARY KEY,
+
+      client_id CHAR(36) NOT NULL,
+      team_id INT UNSIGNED NOT NULL,        /* enroll means user becomes a team member of one challenge */
+      challenge_id INT UNSIGNED NOT NULL,	 
+      url_token CHAR(12) NOT NULL,          /* no foreign key constraints because team's url might change */
```

Comment:
It seems that I misunderstood this. I thought url_token of enrollments comes from the url_token of teams.
Another invited_by is good, tracking who invites who. And let DB generate it on every insertion of enrollments.
But what about the first enrollment url_token within every team? I am thinking it could be anything, and check if the user is the first enrollments of the team. First enrollment of a team with non-matching url_token gets accepted, other enrollments of a team with non-matching url_token is rejected. Should enrollment rejection get some sort of notice, like a pop up modal?

Reply:
That's fine! `invited_by` will probably be `NULL` for most participants. It's just an optional field to track if they signed up with the url `?challenge=pilot&team=<team_url_token>&invite=<someone_elses_enrollment_url_token>`.

Reply:
`url_token` is not optional, though. We need to fill it for each enrollment. I like your idea from below of using `UUID()`!